### PR TITLE
ARROW-14070: [C++][CI] Remove support for VS2015

### PR DESCRIPTION
**Pull request by bkietz**
VS2015 is buggy and Microsoft has withdrawn documentation/bug reporting so it's time for us to drop support.

See mailing list discussion: https://lists.apache.org/thread.html/rff3d8e837aee2964b49733b39a34cffef7c5438dbf2e5c929e878fdf%40%3Cdev.arrow.apache.org%3E

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -55,9 +55,6 @@ environment:
   matrix:
     # NOTE: clcache seems to work best with Ninja and worst with msbuild
     # (as generated by cmake)
-    - JOB: "Build"
-      GENERATOR: Ninja
-      APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2015
```
Comment (pitrou) [inline appveyor.yml:60]: Can we keep this one but switch to a newer VS? It also differs by not using conda packages for dependencies.

File: ci/appveyor-cpp-setup.bat [hunk 2]
```diff
@@ -83,12 +83,11 @@ if "%JOB%" NEQ "Build_Debug" (
...
 if "%GENERATOR%"=="Ninja" set need_vcvarsall=1
 if defined need_vcvarsall (
-    @rem Select desired compiler version
-    if "%APPVEYOR_BUILD_WORKER_IMAGE%" == "Visual Studio 2017" (
-        call "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvarsall.bat" amd64
-    ) else (
-        call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" amd64
+    if "%APPVEYOR_BUILD_WORKER_IMAGE%" != "Visual Studio 2017" (
+        @rem ARROW-14070 Visual Studio 2015 no longer supported
+        exit /B
```
Comment (pitrou) [inline ci/appveyor-cpp-setup.bat:88]: Is this useful?

File: cpp/src/arrow/compute/exec/sink_node.cc [hunk 3]
```diff
@@ -64,22 +64,7 @@ class SinkNode : public ExecNode {
...
-      struct ConvertibleToFuture {
-        operator Future<util::optional<ExecBatch>>() && {  // NOLINT runtime/explicit
-          return std::move(ret);
-        }
-        Future<util::optional<ExecBatch>> ret;
-      };
-
-      return ConvertibleToFuture{push_gen()};
-    };
+    *out_gen = std::move(push_gen);
```
Comment (pitrou) [inline cpp/src/arrow/compute/exec/sink_node.cc:67]: Ah, nice.

**Review by pitrou — COMMENTED** [review]
Thanks for doing this :-)

**Comment by pitrou** [comment]
@github-actions crossbow submit *win*

File: appveyor.yml (hunk 1 above)
Comment (bkietz) [reply appveyor.yml:60]: I've created https://issues.apache.org/jira/browse/ARROW-14075 to add vs2019 and non-conda jobs

**Comment by kszucs** [comment]
The verification tasks require additional prams and the vcpkg task is failing regardless of the PR (based on the nightly report).

**Review by kou — COMMENTED** [review]
There seems to be some issues but overall I like this.

File: cpp/src/gandiva/precompiled/CMakeLists.txt [hunk 4]
```diff
@@ -31,14 +31,13 @@ set(PRECOMPILED_SRCS
...
 
 if(MSVC)
-  # clang pretends to be a particular version of MSVC. Version 1900 is
-  # Visual Studio 2015, 191[0-9] is Visual Studio 2017, and the standard
-  # library uses C++14 features, so we have to use that -std version to
-  # get the IR compilation to work
+  # clang pretends to be a particular version of MSVC. 191[0-9] is
+  # Visual Studio 2017, and the standard library uses C++14 features,
+  # so we have to use that -std version to get the IR compilation to work
   if(MSVC_VERSION MATCHES "^191[0-9]$")
```
Comment (kszucs) [inline cpp/src/gandiva/precompiled/CMakeLists.txt:37]: Does this mean that we support all MSVC versions starting from VS2017?

**Review by kszucs — APPROVED** [review]
LGTM

File: ci/appveyor-cpp-setup.bat (hunk 2 above)
Comment (pitrou) [reply ci/appveyor-cpp-setup.bat:88]: I think you need `exit /B 1` for that.

File: cpp/src/gandiva/precompiled/CMakeLists.txt (hunk 4 above)
Comment (pitrou) [reply cpp/src/gandiva/precompiled/CMakeLists.txt:37]: As much as possible, yes.

**Comment by pitrou** [comment]
Rebased, will merge if CI green.

**Pull request closed without merging**